### ISSUE TITLE
False positive on boxed_local for Box<str> parameters

### ISSUE THREAD

**Issue by BratSinot** [rust-lang/rust-clippy#9971]
### Summary

Sometimes `clippy::boxed_local` gives false-positive to boxed str.

### Lint Name

boxed_local

### Reproducer

I tried this code:

```rust
struct Data {
    msg: [char; 2],
}

impl Data {
    pub fn parse(str: Box<str>) -> Data {
        let mut chars = str.chars();
        Data {
            msg: [chars.next().unwrap(), chars.next().unwrap()],
        }
    }
}

fn main() {
    let data = Data::parse("asd".into());
    let _ = data.msg;
}
```

And got message:
```log
warning: local variable doesn't need to be boxed here
 --> src/main.rs:6:18
  |
6 |     pub fn parse(str: Box<str>) -> Data {
  |                  ^^^
  |
  = note: `#[warn(clippy::boxed_local)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#boxed_local
```

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-apple-darwin
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by Jarcho** [comment]
I don't see the problem with the suggestion here. Taking `&str` instead of `Box<str>` works fine.

**Comment by Jarcho** [comment]
Both `&str` and `Box<str>` have the same representation. They're both a pointer to the string's data along with the size of the string. The difference between them is that `&str` borrows the data, and `Box<str>` owns the data.

Your `parse` function doesn't require ownership of the string. All it does is copy the first two characters from the string into an array. Borrowing is enough for this.

The reason for the lint is that creating a `Box<T>` requires allocating memory for the `Box` to own, whereas using a reference allows the function to use memory which is managed by something else.

**Comment by BratSinot** [comment]
> doesn't require ownership

Aa, right, Forgot about that.
Thanks!
